bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/handler/OpenThermGatewayHandler.java

```diff
@@ -0,0 +1,242 @@
+
+            if (gatewayCommand != null && checkConnection()) {
+                connector.sendCommand(gatewayCommand);
+            }
+        }
+    }
+
+    @Override
+    public void connecting() {
+        connecting = true;
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Connecting");
+    }
+
+    @Override
+    public void connected() {
+        connecting = false;
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    @Override
+    public void disconnected() {
+        connecting = false;
+        
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Disconnected");
+
+        // retry connection if disconnect is not explicitly requested
+        if (!explicitDisconnect && config.connectionRetryInterval > 0) {
+            scheduler.schedule(() -> {
+                if (connector != null && !connecting && !connector.isConnected()) {
+                    connect();
+                }
+            }, config.connectionRetryInterval, TimeUnit.SECONDS);
+        }    
+    }
+
+    @Override
+    public void receiveMessage(Message message) {
+        if (DataItemGroup.dataItemGroups.containsKey(message.getID())) {
+            DataItem[] dataItems = DataItemGroup.dataItemGroups.get(message.getID());
+
+            for (DataItem dataItem: dataItems) {
+                String channelId = dataItem.getSubject();
+
+                if (!OpenThermGatewayBindingConstants.SUPPORTED_CHANNEL_IDS.contains(channelId)) {
+                    continue;
+                }
+
+                State state = null;
+
+                switch (dataItem.getDataType()) {
+                    case FLAGS:
+                        state = OnOffType.from(message.getBit(dataItem.getByteType(), dataItem.getBitPos()));
+                        break;
+                    case UINT8:
+                    case UINT16:
+                        state = new DecimalType(message.getUInt(dataItem.getByteType()));
+                        break;
+                    case INT8:
+                    case INT16:
+                        state = new DecimalType(message.getInt(dataItem.getByteType()));
+                        break;
+                    case FLOAT:
+                        float value = message.getFloat();
+                        @Nullable Unit<?> unit = dataItem.getUnit();
+                        state = (unit == null) ? new DecimalType(value) : new QuantityType<>(value, unit);
+                        break;
+                    case DOWTOD:
+                        break;
+                }
+          
+                if (state != null) {
+                    logger.debug("Received update for channel '{}': {}", channelId, state.toFullString());
+                    updateState(channelId, state);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void handleRemoval() {
+        logger.debug("Removing OpenTherm Gateway handler");
+        disconnect();
+        super.handleRemoval();
+    }
```

Comment:
I'm pretty sure that a handler is disposed before it is removed. So I don't think this code is needed.

Reply:
I'm not worried about the handler itself, I'm calling disconnect() on the background thread when it happens.